synapse/storage/databases/main/lock.py

```diff
@@ -0,0 +1,323 @@
+        )
+        # If there is an entry and it has not timed out, then we still hold the lock
+        if (
+            last_renewed_ts is not None
+            and self._clock.time_msec() - _LOCK_TIMEOUT_MS < last_renewed_ts
+        ):
+            return True
+
+        return False
+
+    async def _renew_lock(self, lock_name: str, lock_key: str, token: str) -> None:
+        """Attempt to renew the lock if we still hold it."""
+        await self.db_pool.simple_update(
+            table="worker_locks",
+            keyvalues={
+                "lock_name": lock_name,
+                "lock_key": lock_key,
+                "token": token,
+            },
+            updatevalues={"last_renewed_ts": self._clock.time_msec()},
+            desc="renew_lock",
+        )
+
+        self._live_tokens.pop((lock_name, lock_key), None)
+
+    async def _drop_lock(self, lock_name: str, lock_key: str, token: str) -> None:
+        """Attempt to drop the lock, if we still hold it"""
+        await self.db_pool.simple_delete(
+            table="worker_locks",
+            keyvalues={
+                "lock_name": lock_name,
+                "lock_key": lock_key,
+                "token": token,
+            },
+            desc="drop_lock",
+        )
+
+
+class Lock:
+    """An async context manager that manages an acquired lock, ensuring it is
+    regularly renewed and dropping it when the context manager exits.
+
+    The context manager returns an async function that can be called to check if
+    the lock is still valid, which can be used to bail out of processing if e.g.
+    processing work in a loop.
+
+    For example:
+
+        lock = await self.store.try_acquire_lock(...)
+        if not lock:
+            return
+
+        async with lock() as still_valid:
+            for item in work:
+                await process(item)
+
+                if not await still_valid():
+                    break
+    """
+
+    def __init__(
+        self,
+        reactor: IReactorCore,
+        clock: Clock,
+        store: LockStore,
+        lock_name: str,
+        lock_key: str,
+        token: str,
+    ) -> None:
+        self._reactor = reactor
+        self._clock = clock
+        self._store = store
+        self._lock_name = lock_name
+        self._lock_key = lock_key
+
+        self._token = token
+
+        self._looping_call = clock.looping_call(self._renew, _RENEWAL_INTERVAL_MS)
+
+        self._dropped = False
+
+    @wrap_as_background_process("Lock._renew")
+    async def _renew(self) -> None:
+        await self._store._renew_lock(self._lock_name, self._lock_key, self._token)
+
+    async def is_lock_still_valid(self) -> bool:
+        """Check if the lock is still held by us"""
+        return await self._store._is_lock_still_valid(
+            self._lock_name, self._lock_key, self._token
+        )
+
+    async def __aenter__(self) -> Callable[[], Awaitable[bool]]:
+        return self.is_lock_still_valid
```

Comment:
maybe assert that we have not dropped the lock? (ie, that we are not reusing the context manager)

Reply:
good point